File: level1/lib/detection/face/face_engine.dart
import '../../core/models.dart';

/// A detected face: bounding box (top-left x/y, bottom-right x/y) + score.
typedef FaceBox = (double, double, double, double);

/// A detected face: bounding box (top-left, bottom-right) + detector score.
class FaceDetection {
  final FaceBox box;
  final double score;

  const FaceDetection({required this.box, required this.score});
}

/// Abstraction over an on-device face detector. Real impl: [TfliteFaceEngine];
/// headless/desktop tests use [MockFaceEngine].
abstract class FaceEngine {
  Future<void> init();

  /// Returns detected faces in [frame]'s color bitmap. Empty list = no faces.
  Future<List<FaceDetection>> detectFaces(ColorBitmap frame);

  Future<void> dispose();
}

File: level1/android/app/src/main/kotlin/io/securitycam/level1/detection/face/FaceEngine.kt
package io.securitycam.level1.detection.face

import io.securitycam.level1.detection.ColorBitmap

/** A detected face: bounding box (top-left x/y, bottom-right x/y) + score. */
data class FaceDetection(
    val x1: Double,
    val y1: Double,
    val x2: Double,
    val y2: Double,
    val score: Double,
)

/** Abstraction over an on-device face detector (port of `face_engine.dart`). */
interface FaceEngine {
    suspend fun init()

    /** Returns detected faces in [frame]'s color bitmap. Empty list = no faces. */
    suspend fun detectFaces(frame: ColorBitmap): List<FaceDetection>

    suspend fun dispose()
}

/** Test/dry-run engine: returns whatever [faces] was pre-loaded with. */
class MockFaceEngine : FaceEngine {
    val faces = mutableListOf<FaceDetection>()

    override suspend fun init() {}

    override suspend fun detectFaces(frame: ColorBitmap): List<FaceDetection> = faces.toList()

    override suspend fun dispose() {}
}